File: python/problems/leetcode/0025_reverse_nodes_in_k_group/solution.py
"""LeetCode 25: Reverse Nodes in k-Group."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def reverseKGroup(
        self, head: Optional[ListNode], k: int
    ) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        pre: ListNode = dummy
        while True:
            tail: Optional[ListNode] = pre
            for _ in range(k):
                tail = tail.next if tail is not None else None
                if tail is None:
                    return dummy.next
            grp_head: Optional[ListNode] = pre.next
            next_grp: Optional[ListNode] = tail.next
            tail.next = None
            new_head, new_tail = self._reverse(grp_head)
            assert new_tail is not None
            pre.next = new_head
            new_tail.next = next_grp
            pre = new_tail

    @staticmethod
    def _reverse(
        head: Optional[ListNode],
    ) -> tuple[Optional[ListNode], Optional[ListNode]]:
        prev: Optional[ListNode] = None
        cur = head
        while cur is not None:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        return prev, head


def from_list(vs: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur = dummy
    for v in vs:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(h: Optional[ListNode]) -> list[int]:
    out: list[int] = []
    while h is not None:
        out.append(h.val)
        h = h.next
    return out


if __name__ == "__main__":
    assert to_list(Solution().reverseKGroup(from_list([1, 2, 3, 4, 5]), 2)) == [
        2, 1, 4, 3, 5,
    ]
    assert to_list(Solution().reverseKGroup(from_list([1, 2, 3, 4, 5]), 3)) == [
        3, 2, 1, 4, 5,
    ]
    assert to_list(Solution().reverseKGroup(from_list([1, 2]), 1)) == [1, 2]
    print("0025 OK")

File: cpp/problems/leetcode/0025_reverse_nodes_in_k_group/solution.cpp
// LeetCode 25
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* n) : val(x), next(n) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode dummy(0, head);
        ListNode* pre = &dummy;
        while (true) {
            ListNode* tail = pre;
            for (int i = 0; i < k; ++i) {
                tail = tail->next;
                if (!tail) return dummy.next;
            }
            ListNode* grp_head = pre->next;
            ListNode* next_grp = tail->next;
            tail->next = nullptr;
            auto [nh, nt] = reverseList(grp_head);
            pre->next = nh;
            nt->next = next_grp;
            pre = nt;
        }
    }

private:
    pair<ListNode*, ListNode*> reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* cur = head;
        while (cur) {
            ListNode* nxt = cur->next;
            cur->next = prev;
            prev = cur;
            cur = nxt;
        }
        return {prev, head};
    }
};

static ListNode* fromList(vector<int> vs) {
    ListNode dummy;
    ListNode* cur = &dummy;
    for (int v : vs) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

static vector<int> toList(ListNode* h) {
    vector<int> out;
    while (h) {
        out.push_back(h->val);
        h = h->next;
    }
    return out;
}

int main() {
    assert((toList(Solution().reverseKGroup(fromList({1, 2, 3, 4, 5}), 2)) ==
            vector<int>{2, 1, 4, 3, 5}));
    assert((toList(Solution().reverseKGroup(fromList({1, 2, 3, 4, 5}), 3)) ==
            vector<int>{3, 2, 1, 4, 5}));
    assert((toList(Solution().reverseKGroup(fromList({1, 2}), 1)) ==
            vector<int>{1, 2}));
    cout << "0025 OK" << endl;
    return 0;
}
